docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,643 @@
+
+```{code-cell} ipython3
+:id: 75FBANKFbmQ5
+
+def matmul_kernel(x_ref, y_ref, z_ref):
+  @pl.when(pl.program_id(2) == 0)
+  def _():
+    z_ref[...] = jnp.zeros_like(z_ref)
+
+  z_ref[...] += x_ref[...] @ y_ref[...]
+
+def matmul(
+    x: jax.Array,
+    y: jax.Array,
+    *,
+    bm: int = 128,
+    bk: int = 128,
+    bn: int = 128,
+):
+  m, k = x.shape
+  _, n = y.shape
+  return pl.pallas_call(
+      matmul_kernel,
+      out_shape=jax.ShapeDtypeStruct((m, n), x.dtype),
+      in_specs=[pl.BlockSpec(lambda i, j, k: (i, k), (bm, bk)),
+                pl.BlockSpec(lambda i, j, k: (k, j), (bk, bn))],
+      out_specs=pl.BlockSpec(lambda i, j, k: (i, j), (bm, bn)),
+      grid=(m // bm, n // bn, k // bk)
+  )(x, y)
+```
+
+```{code-cell} ipython3
+:id: 0e8qTsimccGV
+
+m, k, n = 4096, 4096, 4096
+k1, k2 = random.split(random.key(0), 2)
+x = random.normal(k1, (m, k), dtype=jnp.float32)
+y = random.normal(k2, (k, n), dtype=jnp.float32)
+np.testing.assert_array_equal(x @ y, matmul(x, y))
+```
+
++++ {"id": "DycJX_-PJnnB"}
+
+## Matrix multiplication performance
+
+Let's think about how to analyze matrix multiplication performance.
+
+The number of FLOPs in a `(m, k) x (k, n)` matrix multiplication are (approximately) `2 * m * k * n`. (Technically it is `n * m * (2k - 1)` but for large enough `k` our approximation is sufficient.)
+
+The minimum amount of memory usage for a matrix multiply (assuming float32) is the total size of the inputs (copying into VMEM) plus the size of the output (copying into HBM). Thus the minimum bandwidth usage is `(m * k + k * n + m * n) * 4 bytes/float32`.
+
+One observation is that the number of matmul FLOPs is cubic in its inputs whereas the minimum bandwidth usage is quadratic in its inputs. Intuitively, this means that FLOPs grow faster than bandwidth usage, meaning that the bigger our matmul is, the more compute we have relative to copying.
+
+```{code-cell} ipython3
+:id: HZwmYZ61QZ5L
+:outputId: 18505741-9254-4738-ec64-1660f6733d77
+
+def matmul_flops(m: int, k: int, n: int):
+  return 2 * m * k * n
+
+def matmul_membw(m: int, k: int, n: int, dtype: jnp.dtype):
+  return (m * k + k * n + m * n) * np.dtype(dtype).itemsize
+
+print(matmul_flops(1024, 1024, 1024))
+print(matmul_membw(1024, 1024, 1024, jnp.float32))
+```
+
++++ {"id": "agCtb2GMQazl"}
+
+Now that we can calculate the amount of FLOPs and memory bandwidth usage of a matrix multiplication, let's now see how many FLOP/s we can execute and how much memory bandwidth we actually have on a particular TPU.
+
+This notebook was run on TPU v5e so we will utilize numbers for that specific chip (if you are running this notebook, your numbers may differ). TPU v5es have [197 TFLOPs of bf16 compute and 819 GBps of memory bandwidth](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu_v5e). We can compute the arithmetic intensity that the chip is capable of by taking their ratio (about 240 on TPU v5e).
```

Comment:
GBps -> GB/s?

Reply:
You never explained arithmetic intensity before (just say it's ratio of flop to memory). And in fact it's not enough to just call it "arithmatic intensity". It is the minimal arithmetic intensity a workload must have to not be memory bound in this model.